### ISSUE TITLE
Nullish error case handling

### ISSUE THREAD

**Issue by kosich** [rx-store/rx-store#71]
Hey!

A hypothetical situation: a stream in the `useSubscription` hook might emit a nullish error value

E.g:

```js
const [next, error, complete] = useSubscription(
  throwError(null)
);

// error == null
```

User won't be able to detect this case due to this direct setting error value on state: https://github.com/rx-store/rx-store/blob/master/libs/react/src/lib/react.tsx#L170

**Not sure you need to handle that. Also it might be irrelevant, depending on the store error handling/propagation logic**

**Comment by joshribakoff** [comment]
Yeah right now effects can’t put errors into subjects, because sinks don’t expose it, but it’s possible in react event handlers and possible outside the flow of the store. Good call out!

One idea I’m thinking about is throwing the error in the hook, so it works with error boundaries. One tradeoffs would be that forgetting to handle the error is impossible, it would default to unmounting the tree which is the react and rxjs way! I’m not sure if there’s be downsides or what the other tradeoffs would be

I’m also thinking it could be nice to allow effects to create errors on the subjects too, so effects can trigger the error boundaries.

Maybe one downside is if we put an error on the subject, it can no longer accept any new values... but I think maybe it’s actually a benefit and not a downside to err on the side of safety, and consumers can always pipe catchError or create consuming hooks that swallow the error if they’re so inclined

Very much open to more details on this suggestion and/or PR(s) that would improve it or give more configurable behavior here ⚙️ 

Thanks for taking the time to review and identify this tradeoff 🥳

**Comment by kosich** [comment]
I think, it's a hard question of making stream discrete. I'd agree that propagating error, rather then swallowing it, is better in long term. But then the API becomes a bit strange:

```js
const [next, complete] = useSubscription(stream$);
```

Alternatively you can allow additional callbacks/mappers via similar API that will alter default behavior:

```js
const [next, , complete] = useSubscription(stream$, { error: e => throw e; });
```

Also, an interesting approach is taken in https://observable-hooks.js.org/api/#usesubscription — it accepts `next`, `error`, `complete` handlers. So users decide themselves how to handle stream events. Something like:

```js
const [next, setNext] = useState();
const [complete, setComplete] = useState();
useSubscription(stream$, setNext, e => throw e, setComplete);
```

The error handler could default to `e => throw e` — so that error won't be swallowed. I personally like this approach most.

_The question of swallowing/propagating errors in the store branches is quite fundamental. I think propagating it by default is a good path. Alternatives need more time thinking/playing around._

**Comment by joshribakoff** [comment]
With callbacks is there any need to take them in with the hook's arguments, vs. for example letting users do:

```ts
useSubscription(stream$.pipe(tap(setMyState), catchError(...), finalize(...))
```

If the user wants to handle the `null` error, they could do this, right? I'm assuming in most cases though, the user wants to make their render output based on the last known state, not call a callback. If they did want to call a callback, they can `pipe` `tap` like I'm showing here, or they can use `useEffect` to run a callback when the value of the `error` state changes.

But you do raise a valid point about returning the error vs throwing it, and a good philosophical question about whether errors should err on the side of "safely" unmounting the rest of the app.

Given that this is a React binding for making RxJS work within React nicely, I also agree it is a good idea to evaluate any decision about this in the context of what will be "the react way". I mentioned `ErrorBoundary` in my previous post, but I was actually *wrong* (I like when I get to admit that!). The `ErrorBoundary` is not applicable for errors happening outside of render phase, so any thrown error will not work with an `ErrorBoundary`, I don't think. If throwing the error doesn't work with the `ErrorBoundary`, it seems like it could end up being less ergonomic to try to `catch` the `thrown` error and vary the rendered output based on the presence of an error. In fact, it almost seems like it would be easier (or more ergonomic) to continue to return the `error` state, and then have the user `throw` directly in the render method if they want an `ErrorBoundary`.

To solve the case of needing to be able to handle when a user throws a `null` error on their stream, they can always opt out of the hook's paradigm, and use operators to supply callback like I've shown here? Let me know if you think that's elegant or not.

Assuming you agree with all of this, I can leave this issue open & triage it as a "docs" issue, with the acceptance criteria being to have docs that explain all of this & why we made these tradeoffs. Do you think it also makes sense to track a separate arc of work to make the Rx Store effects forward errors & completions onto the `sinks` (subjects)? Right now adding docs about error handling could cause some confusion, given that the streams will never see an error w/ the current implementation

**Comment by joshribakoff** [comment]
Following up on my error boundary thoughts, another idea is we check the `error` state inside of the hook & throw from within the render phase, but I still think we need `error` as state somewhere even if its internally inside the hook (if we want to use error boundary I think we have to throw in render phase specifically), which still raises the question about the `null` error... for this we can use a special internal JS symbol? This would accomplish the ability to use an error boundary (or would even force users to use one, for better or worse, if the hook throws by default), and unless end users fork the library & export the special JS symbol, they're not going to be able to throw the sentinel value in their app...

**Comment by kosich** [comment]
You're making all the right points with RxJS operators alternative! Again, it's hypothetical and might not come up in real world apps. Keeping that in mind (my intention was to highlight this) and maybe mentioning it in the docs is quite enough for now, IMHO.

> Following up on my error boundary thoughts

Here's how I'm [doing it atm](https://github.com/kosich/react-rxjs-elements/blob/master/src/fragment.ts#L57):

```js
const [error, setError] = useState(null);

// subscription:
.subscribe({
  //...
  error: error => setError({ error })
});

// then in render
if (error) {
  throw error.error;
}
```

Alternatively, you can have two flags for (`hasError` and `errorValue`).

Also note that nullish `next` values suffer of a similar issue, and are also indistinguishable from NEVER stream.

> Do you think it also makes sense to track a separate arc of work to make the Rx Store effects forward errors & completions onto the sinks (subjects)?

I generally think that the core should be treated separately from the react/angular/vue/recks (😁) bridges/adaptors and it's "philosophy" would drive particular APIs in the bridges.

My dev gut tells me that effect errors should be propagated, killing store branches.
In my redux-like rxjs-based store that I did to compliment recks framework experiment — I intentionally propagated all errors from reducers and effects up to the store: have an error — handle it.

_Ofc, all my thoughts here are just suggestions/opinions and it's up to you to decide how to navigate the library further 🔭_

**Comment by joshribakoff** [comment]
Yeah, for the effects I totally agree, to clarify - if an inner effect errors, the outer effect (that spawned it) will also get the error & be killed.

However if the `root` effect used `catchError` on each inner effect it spawned, one child cannot currently put an error onto a subject, this could change though, I also agree it makes sense to start a separate issue about that part & just focus on the React part for this issue.

> nullish next values suffer of a similar issue, and are also indistinguishable from NEVER stream.

Well, we can `.pipe(tap(value=>...))` before subscribing (will want to wrap that in `useMemo` as well) and then it becomes callback based & its no longer an issue... since it is sort of an edge case with a "known workaround", I think it should be a docs change, (if you agree). It seems to me like 99% of users will want to render the value, and maybe 1% of the time you'll need the callback (if you need a callback you can also subscribe directly in `useEffect`, another great thing about this pattern is it has so many escape hatches)

I think we're both sort of on the same page the `error` should be stored in state (perhaps w/ two states like you said), I guess the only remaining open question is whether this hook should `throw` that error automatically & by default. If it did, it would definitely make rx store feel like a more mature & safe pattern, however then the API becomes weird with just `const [next,,complete]` like you said, it feels like if we get rid of `error` maybe we should also get rid of `complete`, and if the user does need to handle completions, they can always use `finalize()`. I'm not quite sure I have a good use case for using `finalize()` on the subjects, and given we do have a good use case for throwing the errors, maybe the API should undergo a breaking change & become:

```
const next = useSubscription()
```

- errors will be stored in state w/ a separate boolean flag about whether it has an error
- errors will be thrown on render, which effectively "kills" the react tree up to nearest error boundary
- no more way to handle `completion` directly
- docs added about supplying callbacks to `catchError`, `tap`, and `finalize` by deriving a new stream in `useMemo`

Sounds good? Let me know if you want to volunteer for any of the PRs 🐣  I definitely would love as many suggestions & contributions as possible, I don't want it to be all up to me, the best open source projects are community driven!

**Comment by kosich** [comment]
Yeah, I was considering this option too.
I think, it should be quite enough for most cases!

> Let me know if you want to volunteer for any of the PRs 🐣

I'm not yet a library user so I'm only loosely familiar with the API / user flows.
Though would love to contribute code-wise later!
